remote_attestation/rust/src/handshaker.rs

```diff
@@ -21,203 +21,153 @@
 // - [`Server`] -> [`Client`]: [`ServerIdentity`]
 // - [`Client`] -> [`Server`]: [`ClientIdentity`]
 //
-// After the protocol handshake both sides create [`AeadEncryptor`] for exchanging encrypted
-// messages.
+// After the protocol handshake both sides create [`Encryptor`] for exchanging encrypted messages.
 
 use crate::{
     crypto::{
         get_random, get_sha256, AeadEncryptor, KeyNegotiator, KeyNegotiatorType, SignatureVerifier,
-        Signer, KEY_AGREEMENT_ALGORITHM_KEY_LENGTH, SIGNING_ALGORITHM_KEY_LENGTH,
+        Signer, SIGNING_ALGORITHM_KEY_LENGTH,
     },
     message::{
-        ClientHello, ClientIdentity, Deserializable, Serializable, ServerIdentity,
-        PROTOCOL_VERSION, REPLAY_PROTECTION_ARRAY_LENGTH,
+        deserialize, ClientHello, ClientIdentity, MessageWrapper, Serializable, ServerIdentity,
     },
     proto::{AttestationInfo, AttestationReport},
 };
 use anyhow::{anyhow, Context};
 use prost::Message;
 
-/// Client of the Remote attestation protocol handshake.
-pub struct ClientAttestationEngine<S>
-where
-    S: AttestationState,
-{
-    /// Behavior of the remote attestation protocol.
-    behavior: AttestationBehavior,
-    /// Current state of the remote attestation protocol.
-    state: S,
-    /// Collection of previously sent and received messages.
-    /// Signed transcript is sent in messages to prevent replay attacks.
-    transcript: Transcript,
+enum HandshakerState {
+    /// Additional state that represents ongoing message processing.
+    MessageProcessing,
+    ExpectingClientHello,
+    ExpectingServerIdentity(KeyNegotiator),
+    ExpectingClientIdentity(KeyNegotiator),
+    Completed(AeadEncryptor),
 }
 
-/// Server of the Remote attestation protocol handshake.
-pub struct ServerAttestationEngine<S>
-where
-    S: AttestationState,
-{
+impl Default for HandshakerState {
+    fn default() -> Self {
+        Self::MessageProcessing
+    }
+}
+
+impl std::fmt::Display for HandshakerState {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        match self {
+            Self::MessageProcessing => write!(f, "MessageProcessing"),
+            Self::ExpectingClientHello => write!(f, "ExpectingClientHello"),
+            Self::ExpectingServerIdentity(_) => write!(f, "ExpectingServerIdentity"),
+            Self::ExpectingClientIdentity(_) => write!(f, "ExpectingClientIdentity"),
+            Self::Completed(_) => write!(f, "Completed"),
+        }
+    }
+}
+
+pub trait Handshaker {
+    fn next_step(&mut self, message: &[u8]) -> anyhow::Result<Option<Vec<u8>>>;
```

Comment:
I am not sure whether there is enough commonality for this extracting this as a shared trait to be very useful. It feels like it is obfuscating details of the process.

Reply:
I also still don't see the value of the additional traits. Encapsulation is done by having pub or private methods, you don't need a trait for that, unless you expect more than one implementation. I guess in this case the two implementation are client and server handshakers, but somehow that still seems backwards to me. Anyway, we can do another pass in the future.

Reply:
Removed the trait